flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSplitMonitoringFunction.java

```diff
@@ -0,0 +1,259 @@
+	}
+
+	@Override
+	public void run(SourceFunction.SourceContext<FileInputSplit> context) throws Exception {
+		FileSystem fileSystem = FileSystem.get(new URI(path));
+		while (isRunning) {
+			monitor(fileSystem, context);
+			Thread.sleep(interval);
+		}
+	}
+
+	private void monitor(FileSystem fs, SourceContext<FileInputSplit> context) throws IOException, JobException {
+		List<FileStatus> files = listEligibleFiles(fs);
+		for (FileInputSplit split : getInputSplitsToForward(files)) {
+			processSplit(split, context);
+		}
+	}
+
+	private void processSplit(FileInputSplit split, SourceContext<FileInputSplit> context) {
+		if (watchType == WatchType.REPROCESS_WITH_APPENDED) {
+			context.collect(split);
+		} else {
+			throw new RuntimeException("Unknown WatchType" + watchType);
+		}
+	}
+
+	private Set<FileInputSplit> getInputSplitsToForward(List<FileStatus> files) throws JobException {
+		if(files.isEmpty()) {
+			return new HashSet<>();
+		}
+
+		FileInputSplit[] inputSplits = createInputSplits();
+
+		Set<FileInputSplit> splitsToForward = new HashSet<>();
+		for (FileInputSplit split: inputSplits) {
+			for(FileStatus file: files) {
+				if (file.getPath().equals(split.getPath())) {
+					splitsToForward.add(split);
+					break;
+				}
+			}
+		}
+		return splitsToForward;
+	}
+
+	/**
+	 * Returns the files that have data to be processed. This method returns the
+	 * Paths to the aforementioned files. It is up to the {@link #processSplit(FileInputSplit, SourceContext)}
+	 * method to decide which parts of the file to be processed, and forward them downstream.
+	 */
+	private List<FileStatus> listEligibleFiles(FileSystem fileSystem) throws IOException {
+		List<FileStatus> files = new ArrayList<>();
+
+		FileStatus[] statuses = fileSystem.listStatus(new Path(path));
+		long maxModificationTime = Long.MIN_VALUE;
+		if (statuses == null) {
+			LOG.warn("Path does not exist: {}", path);
+		} else {
+			// handle the new files
+			for (FileStatus status : statuses) {
+				Path filePath = status.getPath();
+				long modificationTime = status.getModificationTime();
+				if (!shouldIgnore(filePath, modificationTime)) {
+					files.add(status);
+					if(modificationTime > maxModificationTime) {
+						maxModificationTime = modificationTime;
+					}
+				}
+			}
+		}
+
+		// after finding the eligible files to process, update the
+		// global max modification time seen so far to reflect the
+		// latest monitoring round.
+		if(maxModificationTime > globalModificationTime) {
+			globalModificationTime = maxModificationTime;
+		}
+
+		return files;
+	}
+
+	/**
+	 * Returns {@code true} if the file is NOT to be processed further.
+	 * This happens in the following cases:
+	 *
+	 * If the user-specified path filtering method returns {@code true} for the file,
+	 * or if the modification time of the file is smaller than the {@link #globalModificationTime}, which
+	 * signals the point in time where we last checked the path for new data.
```

Comment:
```[...] the {@link #globalModificationTime} [...] signals the point in time where we last checked the path for new data```

this is inaccurate. we can go through a full invokation of listEligibleFiles without changing the globalModificationTime; when no file was modified. The globalModificationTime is the time of the most recent modification found in _any_ of the monitored files.